### DISCUSSION TITLE
General configuration for environment variables

### DISCUSSION THREAD

**Discussion by uilianries** [android/ndk#2011] [Q&A] [answered]
When preparing a script to set then environment variables to be able use Android NDK correctly for a general scenario, without risk of mixing with my native system tool, I would have:

```
CC=/path/to/android/ndk/bin/clang
CXX=/path/to/android/ndk/bin/clang++
LD=/path/to/android/ndk/bin/ld
AR=/path/to/android/ndk/bin/llvm-ar
NM=/path/to/android/ndk/bin/llvm-nm
STRIP=/path/to/android/ndk/bin/llvm-strip
RANLIB=/path/to/android/ndk/bin/llvm-ranlib
OBJCOPY=/path/to/android/ndk/bin/llvm-objcopy
OBJDUMP=/path/to/android/ndk/bin/llvm-objdump
READELF=/path/to/android/ndk/bin/llvm-readelf
ADDR2LINE=/path/to/android/ndk/bin/llvm-addr2line
```

However, when reading the recommendation from the Build System Maintainers Guide

- https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#clang
- https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#linkers
- https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#binutils

They said:

> Note: It is usually not necessary to invoke the linkers directly since Clang will do so automatically. Clang will also automatically link CRT objects and default libraries and set up other target-specific options, so it is generally better to use Clang for linking.

And

> Note that llvm-as is not an equivalent of GNU as, but rather a tool for assembling LLVM IR. If you are currently using as directly, you will need to migrate to using clang as a driver for building assembly.

So it seems like I don't need to specify all these variables, only `CC` and `CXX`, is that correct? 

However, https://developer.android.com/ndk/guides/other_build_systems#autoconf points everything listed:

```
# Check out the source.
git clone https://github.com/glennrp/libpng -b v1.6.37
cd libpng
# Only choose one of these, depending on your build machine...
export TOOLCHAIN=$NDK/toolchains/llvm/prebuilt/darwin-x86_64
export TOOLCHAIN=$NDK/toolchains/llvm/prebuilt/linux-x86_64
# Only choose one of these, depending on your device...
export TARGET=aarch64-linux-android
export TARGET=armv7a-linux-androideabi
export TARGET=i686-linux-android
export TARGET=x86_64-linux-android
# Set this to your minSdkVersion.
export API=21
# Configure and build.
export AR=$TOOLCHAIN/bin/llvm-ar
export CC=$TOOLCHAIN/bin/$TARGET$API-clang
export AS=$CC
export CXX=$TOOLCHAIN/bin/$TARGET$API-clang++
export LD=$TOOLCHAIN/bin/ld
export RANLIB=$TOOLCHAIN/bin/llvm-ranlib
export STRIP=$TOOLCHAIN/bin/llvm-strip
./configure --host $TARGET
make
```

So I'm not sure which approach is better/safer.

**Comment by enh-google** [comment, marked as answer]
On Mon, Apr 8, 2024 at 3:42 AM Uilian Ries ***@***.***> wrote:

> When preparing a script to set then environment variables to be able use
> Android NDK correctly for a general scenario, without risk of mixing with
> my native system tool, I would have:
>
> CC=/path/to/android/ndk/bin/clang
> CXX=/path/to/android/ndk/bin/clang++
> LD=/path/to/android/ndk/bin/ld
> AR=/path/to/android/ndk/bin/llvm-ar
> NM=/path/to/android/ndk/bin/llvm-nm
> STRIP=/path/to/android/ndk/bin/llvm-strip
> RANLIB=/path/to/android/ndk/bin/llvm-ranlib
> OBJCOPY=/path/to/android/ndk/bin/llvm-objcopy
> OBJDUMP=/path/to/android/ndk/bin/llvm-objdump
> READELF=/path/to/android/ndk/bin/llvm-readelf
> ADDR2LINE=/path/to/android/ndk/bin/llvm-addr2line
>
> However, when reading the recommendation from the Build System Maintainers
> Guide
>
>    -
>    https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#clang
>    -
>    https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#linkers
>    -
>    https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#binutils
>
> They said:
>
> Note: It is usually not necessary to invoke the linkers directly since
> Clang will do so automatically. Clang will also automatically link CRT
> objects and default libraries and set up other target-specific options, so
> it is generally better to use Clang for linking.
>
> And
>
> Note that llvm-as is not an equivalent of GNU as, but rather a tool for
> assembling LLVM IR. If you are currently using as directly, you will need
> to migrate to using clang as a driver for building assembly.
>
> So it seems like I don't need to specify all these variables, only CC and
> CXX, is that correct?
>
the word "usually" in the docs is important. most builds are simple, and
simple builds will just call "the compiler" to both compile and link stuff.

if, on the other hand, you're dealing with a build that uses niche tools
like ranlib, you're much more likely to want to set all of these.

but in practice, "only you can answer this question" --- look at what your
build uses, and make sure you've set those variables.

(there's nothing magic about these environment variables --- they're
literally just "commonly used environment variables in gnu
autotools/configure style builds.)
